core/modules/qhttp/Path.cc

```diff
@@ -0,0 +1,175 @@
+/*
+ * LSST Data Management System
+ * Copyright 2017 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "Path.h"
+
+// System headers
+#include <string>
+#include <vector>
+
+// Third-party headers
+#include "boost/format.hpp"
+#include "boost/regex.hpp"
+#include "boost/variant.hpp"
+
+namespace {
+
+std::string escapeString(std::string const& str)
+{
+    static boost::regex escapeRe{"([.[{}()\\*+?|^$])"};
+    return regex_replace(str, escapeRe, "\\1");
+}
+
+std::string escapeGroup(std::string const& str)
+{
+    static boost::regex escapeRe{"([=!:$/()])"};
+    return regex_replace(str, escapeRe, "\\1");
+}
+
+struct PathToken
+{
+    std::string name;
+    std::string prefix;
+    std::string delimeter;
+    bool optional;
+    bool repeat;
+    std::string pattern;
+};
+
+} // anon. namespace
+
+namespace lsst {
+namespace qserv {
+namespace qhttp {
+
+
+void Path::parse(std::string const& pattern)
+{
+    static boost::regex patternRe{
+        // Match escaped characters that would otherwise appear in future matches.
+        // This allows the user to escape special characters that won't transform.
+        "(\\\\.)|"
+        // Match Express-style parameters and un-named parameters with a prefix
+        // and optional suffixes. Matches appear as:
+        // "/:test(\\d+)?" => [ "/", "test", "\d+", -----, "?", --- ]
+        // "/route(\\d+)"  => [ ---, ------, -----, "\d+", ---, --- ]
+        // "/*"            => [ "/", ------, -----, -----, ---, "*" ]
+        "([\\/.])?(?:(?:\\:(\\w+)(?:\\(((?:\\\\.|[^()])+)\\))?|\\(((?:\\\\.|[^()])+)\\))([+*?])?|(\\*))"
```

Comment:
Yes, it would be nice to see the raw string and translate it if needed (to make C++ happy) before passing into boost::regex. I also think this method requires an extensive documentation for the regular expression and the way it's used.